src/interface.jl

```diff
@@ -133,10 +135,13 @@ function save!!(
     iteration::Integer,
     ::AbstractModel,
     ::AbstractSampler,
-    ::Integer;
+    N::Integer;
     kwargs...
 )
-    return BangBang.setindex!!(transitions, transition, iteration)
+    @assert length(transitions) == iteration - 1
+    new_ts = BangBang.push!!(transitions, transition)
+    typeof(new_ts) !== typeof(transitions) && Base.sizehint!(new_ts, N)
+    return new_ts
```

Comment:
IMO the default implementation should only `push!!`:
```suggestion
    return BangBang.push!!(transitions, transition)
```

Reply:
`sizehint!` is needed to get similar performance to `setindex!!`.

Reply:
No, that happens once during initialization of `transitions`. There is no need to call `sizehint!` in every iteration.

Reply:
Only if the type changes because the output and input vectors would be different.

Reply:
The type will likely only change a few times so calling `sizehint!` here will only happen a few times.

Reply:
Ah I see, although I'm still not happy that about using `sizehint!` in this general definition. Think about the thinning example above, even if it is handled correctly during initialization we would suddenly preallocate much more memory than needed just because the type changed.

Reply:
So I think maybe rather downstream packages should have to pay the penalty for changing types or provide their custom implementation.

Reply:
Well the allocation was happening previously anyways. If we need to store less samples than `N` then maybe that should change in the `mcmcsample` function not `save!!`.

Reply:
The main idea behind the `save!!` functionality is to allow for a simple way to implement thinning without saving all transitions first or saving results to disk instead of storing them in memory.

But aside from that, I think it's probably not the best example since in that case you would implement `save!!` anyways for your own types. I'm just a bit worried about making any assumptions about how `transitions` is implemented in the default implementation of `save!!`. Users are supposed to be able to just change `transitions` if they want some other container type that maybe does not support `sizehint!`, without having to implement `save!!`. In the default implementation of `transitions` `sizehint!` makes sense since it returns a vector of transitions anyways, but IMO the natural default implementation of ``save!!`` would just use `push!!` - we just don't know if the user preallocated memory or wants to preallocate memory, and if yes, how much. So the simplest thing would be not to try to be clever at all and just use the same default as for the iterator-based sampling.

Reply:
So just because the default implementation uses a vector and `sizehint!` doesn't mean any other implementation should. `push!` may not even be used in the custom implementation. `N` and pre-allocation can be ignored in any implementation where it makes sense to ignore it. But for this specific implementation, it seems reasonable to use `sizehint!`. For thinning, won't you have to implement a different `mcmcsample` or `save!!` anyways? So then we can remove the `sizehint!` in that method, no?

Reply:
> For thinning, won't you have to implement a different mcmcsample or save!! anyways

You would not touch `mcmcsample` but implement a different `save!!` ideally (because otherwise all the default logic is gone). As I said it's not really an appropriate example in this context since you would implement `save!!` anyways.

A better example would be static arrays. With `push!!` you could actually save the transitions in static arrays but `sizehint!` is not defined for static arrays. The same should be true for any array type of fixed size.

If we don't care about breaking these cases for everyone we could do something like
- Define `save!!` only for `AbstractVector` container of transitions in AbstractMCMC
- Check if `newts === transitions` (that will check the types first but also make sure that the arrays are identical, since otherwise the size hint was lost as well probably) and apply `sizehint!` otherwise if the total number of iterations is provided.

However, I assume that depending on how often the type changes and how long the chains are, this could actually be less efficient than just pushing to the new vectors. And if the array type doesn't change the implementation without `sizehint!` would be very efficient since we call `sizehint!` when initializing the container in `transitions`. So I'm not sure there's much benefit in adding `sizehint!` here since it breaks some things that are currently possible and might still not be completely efficient (that would require some benchmarking though).

Reply:
BTW for thinning it would be much nicer to work with transducers or iterators I guess.

Reply:
I don't know why we may need to save the chains a static array. We can use `sizehint!` only for `AbstractVector`. That sounds fair enough.

>Check if newts === transitions (that will check the types first but also make sure that the arrays are identical, since otherwise the size hint was lost as well probably) and apply sizehint! otherwise if the total number of iterations is provided.

If the actual variable is changing but the type is the same, then that would be a bug in BangBang. If the type is not changing, there is no reason why we can't just `push!` normally. Anyways, this will be inside an if statement checking if `AbstractVector` or not. So it should be safe for methods of this function that don't use an `AbstractVector` for some reason. In fact, I would argue that using `BangBang` we are assuming that `transitions` is an instance of `AbstractVector` anyways so we might as well constraint the argument type. Overloading `save!!` for a custom `transitions` type makes more sense to me than overloading `BangBang.push!!`.

Reply:
> If the actual variable is changing but the type is the same, then that would be a bug in BangBang.

Not necessarily, you could have immutable containers whose size is not included in the type information.

I would be fine (but not completely satisfied :stuck_out_tongue:) with the suggestions in https://github.com/TuringLang/AbstractMCMC.jl/pull/40/files#r427928351, I guess it should be fine in most cases. If we experience any issues with `sizehint!` (bugs or performance), we can remove it.